panther_hardware_interfaces/include/panther_hardware_interfaces/gpio_controller.hpp

```diff
@@ -0,0 +1,294 @@
+#include <vector>
+
+#include <gpiod.hpp>
+
+#include "panther_gpiod/gpio_driver.hpp"
+
+namespace panther_hardware_interfaces
+{
+
+class Watchdog
+{
+public:
+  /**
+   * @brief Constructor for Watchdog class.
+   *
+   * @param gpio_driver Pointer to the GPIODriver object.
+   * @exception std::runtime_error if the Watchdog pin is not configured by GPIODriver or not
+   * described in GPIOController gpio_info storage
+   */
+  Watchdog(std::shared_ptr<panther_gpiod::GPIODriver> gpio_driver);
+
+  /**
+   * @brief Destructor for Watchdog class. Turns off the watchdog thread.
+   */
+  ~Watchdog();
+
+  /**
+   * @brief Turns on the Watchdog.
+   *
+   * @return True if successful; otherwise, false.
+   * @exception std::runtime_error if the Watchdog pin is not configured by GPIODriver or not
+   * described in GPIOController gpio_info storage
+   */
+  bool TurnOn();
+  bool TurnOff();
+  bool IsWatchdogEnabled() const;
+
+private:
+  /**
+   * @brief Monitors the Watchdog thread status.
+   *
+   * While the Watchdog thread is enabled, it toggles the state of the Watchdog pin between active
+   * and inactive states at regular intervals. The Watchdog thread is active for a duration of 10
+   * milliseconds and then sleeps for the same duration. When the Watchdog thread is disabled, it
+   * sets the Watchdog pin to an inactive state.
+   */
+  void WatchdogThread();
+
+  panther_gpiod::GPIOPin watchdog_pin_ = panther_gpiod::GPIOPin::WATCHDOG;
+  std::shared_ptr<panther_gpiod::GPIODriver> gpio_driver_;
+  std::unique_ptr<std::thread> watchdog_thread_;
+  std::atomic_bool watchdog_thread_enabled_ = false;
+};
+
+class GPIOControllerInterface
+{
+public:
+  virtual void Start() = 0;
+  virtual bool MotorsEnable(const bool enable) = 0;
+  virtual bool FanEnable(const bool enable) = 0;
+  virtual bool AUXEnable(const bool enable) = 0;
+  virtual bool VDIGEnable(const bool enable) = 0;
+  virtual bool ChargerEnable(const bool enable) = 0;
+  virtual bool EStopTrigger() = 0;
+  virtual bool EStopReset() = 0;
+
+  /**
+   * @brief This method sets the provided callback function to be executed upon GPIO edge events.
+   *
+   * @param callback The callback function to handle GPIO edge events.
+   *
+   * @par Example
+   * An example of using this method to bind a member function as a callback:
+   * @code{.cpp}
+   * class MyClass {
+   * public:
+   *   void HandleGPIOEvent(const GPIOInfo & gpio_info) {
+   *     // Handle GPIO event here, i.e:
+   *     std::cout << gpio_info.offset << ":    " << gpio_info.value << std::endl;
+   *   }
+   * };
+   *
+   * MyClass my_obj;
+   * GPIOControllerPTH12X gpio_controller;
+   * gpio_controller.ConfigureGpioStateCallback(
+   *     std::bind(&MyClass::HandleGPIOEvent, &my_obj, std::placeholders::_1));
+   * @endcode
+   */
+  void ConfigureGpioStateCallback(
+    const std::function<void(const panther_gpiod::GPIOInfo &)> & callback);
+
+  bool IsPinActive(const panther_gpiod::GPIOPin pin) const;
```

Comment:
```suggestion
  bool IsPinActive(const panther_gpiod::GPIOPin & pin) const;
```
Not sure if we should add reference for enum class tho

Reply:
It's similar to an integer. The copy process is faster, and there is no dereference stage when using this parameter in the method.